src/main/java/com/fcgl/xml/CompoundXMLNode.java

```diff
@@ -0,0 +1,193 @@
+    Objects.requireNonNull(node);
+
+    if (node.isElementNode()) {
+      return true;
+    }
+    // Added children
+    List<IXMLNode> children = node.getChildren().orElse(Collections.emptyList());
+    Stack<IXMLNode> nodeToVisit = new Stack<>();
+    children.forEach(nodeToVisit::push);
+
+    // Direct Ancestor, ONLY add IXMLCompound with children
+    Stack<IXMLNode> directAncestor = new Stack<>();
+    directAncestor.push(node);
+    IXMLNode curParent = node;
+
+    List<IXMLNode> visited = new ArrayList<>();
+    visited.add(node);
+    while (!nodeToVisit.empty()) {
+      IXMLNode curNode = nodeToVisit.pop();
+
+      // Add children to the visit stack
+      if (!curNode.isElementNode()
+          && !visited.contains(curNode)
+          && curNode.getChildren().isPresent()) {
+        curNode.getChildren().get().forEach(nodeToVisit::push);
+      }
+
+      if (directAncestor.contains(curNode)) {
+        return false;
+      }
+      visited.add(curNode);
+
+      boolean isChildOfCurrentParent =
+          curParent.getChildren().orElse(Collections.emptyList()).contains(curNode);
+      if (!isChildOfCurrentParent) {
+        directAncestor.pop();
+        curParent = directAncestor.peek();
+      }
+      // Parents without any children, don't have an kids to be ancestor
+      // Only add parents when they have kids
+      if (!curNode.isElementNode()
+          && curNode.getChildren().orElse(Collections.emptyList()).size() > 0) {
+        directAncestor.add(curNode);
+        curParent = curNode;
+      }
+    }
+    return true;
+  }
+
+  private void setNodeXmlTag(Tag tag) {
+    this.nodeXmlTag = tag;
+  }
+
+  private Tag getNodeXmlTag() {
+    return this.nodeXmlTag;
+  }
+  @Override
+  public String getNodeName() {
+    return this.nodeName;
+  }
+
+  @Override
+  public boolean isElementNode() {
+    return false;
+  }
+
+  private List<IXMLNode> getChildrenNode() {
+    return this.childrenNode;
+  }
+
+  private void setChildrenNode(List<IXMLNode> childrenNode) {
+    this.childrenNode = childrenNode;
+  }
+
+  @Override
+  public Optional<List<IXMLNode>> getChildren() {
+    List<IXMLNode> unModifiableList = Collections.unmodifiableList(this.getChildrenNode());
+    return Optional.of(unModifiableList);
+  }
+
+  @Override
+  public void addChildren(IXMLNode node) {
+    Objects.requireNonNull(node);
+
+    List<IXMLNode> originalChilren = this.getChildrenNode();
+    // the same "pointer" object, crosses are ok
+    List<IXMLNode> temp = new LinkedList<>(this.getChildrenNode());
+    temp.add(node);
+    this.setChildrenNode(temp);
+    if (node == this || !isChildValid(this)) {
+      //reset to the original list
+      this.setChildrenNode(originalChilren);
+      throw new IllegalArgumentException("Ancestor can't both be child/grandchild of itself");
+    }
+    this.setChildrenNode(originalChilren);
+    this.getChildrenNode().add(node);
+  }
+
+  @Override
+  public IXMLNode remove(int atIndex) {
```

Comment:
What would the use case for a remove method be? I don't think we need it do we? If you have a good use case and we expect to use it a lot maybe a LinkedList isn't the best data structure for this deletion.

Reply:
One of a few data structure few guarantee order. Usually no, therefore justifiable.

Reply:
Also LinkList remove index is O(1) so can't get faster.

Reply:
removing in LinkedList is O(1) but searching is O(n). So it takes O(n) to remove something because you have to search for it first

Reply:
I am not searching. I removed a specific index.

Reply:
Sorry I wasn't being clear. When you specify a specific index in a LinkedList that the "search". The head node is the only node that can be removed at O(1) time. Anything else is O(n) because it has to traverse the linked list to get to a specific index. 

https://stackoverflow.com/questions/33987542/time-complexity-of-deletion-in-a-linked-list

Anyways we said not to worry about time complexity at the moment.